middleware/admin/certificate_v1.py

```diff
@@ -0,0 +1,251 @@
+    EXTRACTORS = {
+        "device": lambda b: b[-65:],
+        "attestation": lambda b: b[1:],
+        "ui": lambda b: b[:],
+        "signer": lambda b: b[:],
+    }
+
+    def __init__(self, element_map):
+        if ("name" not in element_map
+                or element_map["name"] not in self.VALID_NAMES):
+            raise ValueError("Missing or invalid name for HSM certificate element")
+        self._name = element_map["name"]
+
+        if "signed_by" not in element_map:
+            raise ValueError("Missing certifier for HSM certificate element")
+        self._signed_by = element_map["signed_by"]
+
+        self._tweak = None
+        if "tweak" in element_map:
+            if not is_nonempty_hex_string(element_map["tweak"]):
+                raise ValueError(
+                    f"Invalid signer tweak for HSM certificate element {self.name}")
+            self._tweak = element_map["tweak"]
+
+        if "message" not in element_map or not is_nonempty_hex_string(
+                element_map["message"]):
+            raise ValueError(
+                f"Missing or invalid message for HSM certificate element {self.name}")
+        self._message = element_map["message"]
+
+        if "signature" not in element_map or not is_nonempty_hex_string(
+                element_map["signature"]):
+            raise ValueError(
+                f"Missing or invalid signature for HSM certificate element {self.name}")
+        self._signature = element_map["signature"]
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def signed_by(self):
+        return self._signed_by
+
+    @property
+    def tweak(self):
+        return self._tweak
+
+    @property
+    def message(self):
+        return self._message
+
+    @property
+    def signature(self):
+        return self._signature
+
+    def to_dict(self):
+        result = {
+            "name": self.name,
+            "message": self.message,
+            "signature": self.signature,
+            "signed_by": self.signed_by,
+        }
+
+        if self.tweak is not None:
+            result["tweak"] = self.tweak
+
+        return result
+
+    def is_valid(self, certifier_pubkey):
+        try:
+            message = bytes.fromhex(self.message)
+
+            verifier_pubkey = certifier_pubkey
+            if self.tweak is not None:
+                tweak = hmac.new(
+                    bytes.fromhex(self.tweak),
+                    certifier_pubkey.serialize(compressed=False),
+                    hashlib.sha256,
+                ).digest()
+
+                verifier_pubkey = verifier_pubkey.tweak_add(tweak)
+
+            return verifier_pubkey.ecdsa_verify(
+                message, verifier_pubkey.ecdsa_deserialize(bytes.fromhex(self.signature)))
+        except Exception:
+            return False
+
+    def get_value(self):
+        return self.EXTRACTORS[self.name](bytes.fromhex(self.message)).hex()
+
+
+class HSMCertificate:
```

Comment:
Is there a reason for not moving `HSMCertificate` class to `certificate.py` and keeping the common logic between v1 and v2 there instead?

Then this file would define `HSMCertificateV1` with v1 specific logic, similarly to what is done in `certificate_v2.py`.

Reply:
Didn't think much about it to be honest. I guess just to not have an almost empty `HSMCertificate` class with a single class method. Something to improve in the future. Thanks.